llvm/lib/Target/AIE/AIEWawRegRewriter.cpp

```diff
@@ -0,0 +1,414 @@
+  assert(Reg.isVirtual());
+
+  // Only rename registers mapped to a phys reg assigned more than once
+  if (!UsedPhysRegs[VRM->getPhys(Reg)])
+    return false;
+
+  if (!TRI->isVecOrAccRegClass(*(MRI->getRegClass(Reg))))
+    return false;
+
+  return !VRegWithCopies[Reg.virtRegIndex()];
+}
+
+BitVector
+AIEWawRegRewriter::getDefinedPhysRegs(const MachineBasicBlock *MBB) const {
+  BitVector BlockedPhysRegs(TRI->getNumRegs());
+
+  for (const MachineInstr &MI : *MBB) {
+    for (const MachineOperand &Op : MI.defs()) {
+      MCPhysReg PhysReg = getAssignedPhysReg(Op.getReg());
+      if (MCRegister::isPhysicalRegister(PhysReg))
+        addAliasRegs(BlockedPhysRegs, PhysReg);
+    }
+  }
+
+  return BlockedPhysRegs;
+}
+
+bool AIEWawRegRewriter::replaceReg(const Register Reg,
+                                   BitVector &BlockedPhysRegs) {
+  assert(Reg.isVirtual());
+  LLVM_DEBUG(dbgs() << " WAW RegRewriter: Register to replace "
+                    << TRI->getName(VRM->getPhys(Reg)) << "\n");
+
+  MCPhysReg ReplacementPhysReg = getReplacementPhysReg(Reg, BlockedPhysRegs);
+
+  if (!MCRegister::isPhysicalRegister(ReplacementPhysReg))
+    return false;
+
+  LLVM_DEBUG(dbgs() << "     WAW Replacement: Virtual Register "
+                    << printReg(VRM->getPhys(Reg), TRI, 0, MRI)
+                    << " will replace "
+                    << printReg(VRM->getPhys(Reg), TRI, 0, MRI) << " with "
+                    << printReg(ReplacementPhysReg, TRI, 0, MRI) << '\n');
+
+  const LiveInterval &LI = LIS->getInterval(Reg);
+  LRM->unassign(LI);
+  LRM->assign(LI, ReplacementPhysReg);
+  addAliasRegs(BlockedPhysRegs, ReplacementPhysReg);
+  return true;
+}
+
+MCPhysReg AIEWawRegRewriter::getReplacementPhysReg(
+    const Register Reg, const BitVector &BlockedPhysRegs) const {
+  assert(Reg.isVirtual() && "Reg has to be a virtual register");
+  const TargetRegisterClass *RC = MRI->getRegClass(Reg);
+
+  LiveInterval &LI = LIS->getInterval(Reg);
+  for (const MCPhysReg &PhysReg : RC->getRegisters()) {
+
+    if (BlockedPhysRegs[PhysReg])
+      continue;
+
+    LiveRegMatrix::InterferenceKind IK = LRM->checkInterference(LI, PhysReg);
+    if (IK == LiveRegMatrix::IK_Free)
+      return PhysReg;
+  }
+  return MCRegister::NoRegister;
+}
+
+bool AIEWawRegRewriter::isIdentityCopy(const MachineInstr &MI) const {
+  std::optional<DestSourcePair> DestSource = TII->isCopyInstr(MI);
+  if (!DestSource.has_value())
+    return false;
+
+  return getAssignedPhysReg(DestSource->Source->getReg()) ==
+         getAssignedPhysReg(DestSource->Destination->getReg());
+}
+
+IndexedMap<const MachineInstr *, VirtReg2IndexFunctor>
+AIEWawRegRewriter::getLastVRegDef(const MachineBasicBlock &MBB,
+                                  const int MaxVRegIndex) const {
+  IndexedMap<const MachineInstr *, VirtReg2IndexFunctor> LastVRegDef;
+
+  // Initialize the IndexedMap size
+  if (MaxVRegIndex > 0)
+    LastVRegDef.grow(MaxVRegIndex);
```

Comment:
Nit: I think we could resize that map once and for all using `MRI.getNumVirtRegs()`

Reply:
fixed it